src/compiler/checker.ts

```diff
@@ -5353,54 +5352,58 @@ namespace ts {
                 case SyntaxKind.LiteralType:
                     return true;
                 case SyntaxKind.ArrayType:
-                    return isIndependentType((<ArrayTypeNode>node).elementType);
+                    return isTypeFreeOfThisReference((<ArrayTypeNode>node).elementType);
                 case SyntaxKind.TypeReference:
-                    return isIndependentTypeReference(<TypeReferenceNode>node);
+                    return isTypeReferenceFreeOfThisReference(<TypeReferenceNode>node);
             }
             return false;

```

Comment:
So all other types are presumed guilty? Could you try for a complete `switch` and `Debug.fail()` in the default case?

Reply:
In a word, yes. I'm doing Definitely Typed duty this week, so I ran tests on the assert until they all passed. I had to add all the type kinds except for Mapped Types.

There may be some efficiency we could pick up here by inspecting more types &mdash; union and intersection especially &mdash; but I think it's fine to have the default case fall through to a `return false`.